astropy/convolution/convolve.py

```diff
@@ -14,20 +17,54 @@
 from ..nddata import support_nddata
 from ..modeling.core import _make_arithmetic_operator, BINARY_OPERATORS
 from ..modeling.core import _CompoundModelMeta
-
-
+from .utils import KernelSizeError, has_even_axis, raise_even_kernel_exception
+from ..utils.openmp import handle_n_threads_usage
+
+# Turn the faulthandler ON to help catch any signals, e.g. segfaults
+# or asserts. This doesn't, currently, work with Jupyter Notebook.
+faulthandler.enable()
+
+# Find and load C convolution library
+lib_path = glob.glob(os.path.join(os.path.dirname(__file__), 'lib_convolve*'))[0]
+lib = ctypes.cdll.LoadLibrary(lib_path)
+# The GIL is automatically released by default when calling functions imported
+# from libaries loaded by ctypes.cdll.LoadLibrary(<path>)
+
+# Declare prototypes
+# Boundary None
+_convolveNd_boundary_none_c = lib.convolveNd_boundary_none_c
+_convolveNd_boundary_none_c.restype = None
+_convolveNd_boundary_none_c.argtypes = [ndpointer(ctypes.c_double, flags={"C_CONTIGUOUS", "WRITEABLE"}), # return array
```

Comment:
@astrofrog 

>>The ctypes decl for the C functions conducts a runtime type check. The cython code would need to do this manually (ourselves) or just static cast which is not a good idea.
>
>Could you point to the specific code you mean here?

This is the spot.

Read the [Ctypes argtypes para](https://docs.python.org/3/library/ctypes.html#specifying-the-required-argument-types-function-prototypes) and let me know what you think.

More importantly, here's the `numpy.ctypeslib.ndpointer()` [doc](https://docs.scipy.org/doc/numpy-1.13.0/reference/routines.ctypeslib.html#numpy.ctypeslib.ndpointer) stating:

> Array-checking restype/argtypes.
>
>An ndpointer instance is used to describe an ndarray in restypes and argtypes specifications. This approach is more flexible than using, for example, POINTER(c_double), since several restrictions can be specified, which are verified upon calling the ctypes function. These include data type, number of dimensions, shape and flags. If a given array does not satisfy the specified restrictions, a TypeError is raised.